### ISSUE TITLE
4.0.8 DicomClient SendAsync never returns or throws

### ISSUE THREAD

**Issue by saydo17** [fo-dicom/fo-dicom#1325]
**Describe the bug**
SendAsync never returns when socket times out while reading PDU.

```
022-02-09 11:22:00,212 - [25] - DEBUG - Dicom - [IDLE] No requests to send, staying idle
2022-02-09 11:22:01,938 - [41] - DEBUG - Dicom - [IDLE] --> [CONNECTING]
2022-02-09 11:22:02,048 - [17] - DEBUG - Dicom - [CONNECTING] --> [REQUESTING ASSOCIATION]
2022-02-09 11:22:02,048 - [17] - INFO - Dicom - PACSAE -> Association request:
Calling AE Title:          OEM_StoreSCU
Called AE Title:           PACSAE
Remote host:               192.0.2.1
Remote port:               104
Implementation Class:      Implementation Class UID [1.3.6.1.4.1.30071.8]
Implementation Version:    fo-dicom 4.0.8
Maximum PDU Length:        0
Async Ops Invoked:         0
Async Ops Performed:       0
Presentation Contexts:     1
  Presentation Context:     1 [Proposed]
       Abstract Syntax:     Digital X-Ray Image Storage - For Presentation
       Transfer Syntax:     Implicit VR Little Endian: Default Transfer Syntax for DICOM
2022-02-09 11:22:02,171 - [33] - INFO - Dicom - PACSAE <- Association accept:
Calling AE Title:          OEM_StoreSCU
Called AE Title:           PACS
Remote host:               192.0.2.1
Remote port:               104
Implementation Class:      Unknown [1.2.40.0.13.1.1]
Implementation Version:    dcm4che-2.0
Maximum PDU Length:        16384
Async Ops Invoked:         0
Async Ops Performed:       0
Presentation Contexts:     1
  Presentation Context:     1 [Accept]
       Abstract Syntax:     Digital X-Ray Image Storage - For Presentation
       Transfer Syntax:     Implicit VR Little Endian: Default Transfer Syntax for DICOM
2022-02-09 11:22:02,171 - [10] - DEBUG - Dicom - [REQUESTING ASSOCIATION] Association is accepted
2022-02-09 11:22:02,171 - [10] - DEBUG - Dicom - [REQUESTING ASSOCIATION] --> [SENDING REQUESTS]
2022-02-09 11:22:02,171 - [10] - DEBUG - Dicom - [SENDING REQUESTS] Sending DICOM requests
2022-02-09 11:22:02,172 - [10] - INFO - Dicom - PACSAE -> C-Store request [8]
2022-02-09 11:30:10,877 - [11] - INFO - Dicom - PACSAE -> C-Store request [9]
2022-02-09 11:30:11,814 - [11] - INFO - Dicom - PACSAE <- C-Store response [8]: Success
2022-02-09 11:36:08,533 - [50] - INFO - Dicom - PACSAE -> C-Store request [10]
2022-02-09 11:36:08,931 - [33] - INFO - Dicom - PACSAE <- C-Store response [9]: Success
2022-02-09 11:39:05,485 - [44] - INFO - Dicom - PACSAE -> C-Store request [11]
2022-02-09 11:39:07,029 - [33] - INFO - Dicom - PACSAE <- C-Store response [10]: Success
2022-02-09 11:39:49,964 - [16] - INFO - Dicom - Socket error while reading PDU: TimedOut [10060]
2022-02-09 11:39:49,964 - [35] - INFO - Dicom - Socket error while writing PDU: ConnectionReset [10054]
2022-02-09 11:39:49,964 - [16] - INFO - Dicom - Connection closed
2022-02-09 11:39:49,965 - [46] - INFO - Dicom - Connection closed
2022-02-09 11:39:49,982 - [50] - ERROR - Dicom - Exception creating PDV: System.IO.IOException: Unable to write data to the transport connection: An existing connection was forcibly closed by the remote host. ---> System.Net.Sockets.SocketException: An existing connection was forcibly closed by the remote host
   at System.Net.Sockets.Socket.EndSend(IAsyncResult asyncResult)
   at System.Net.Sockets.NetworkStream.EndWrite(IAsyncResult asyncResult)
   --- End of inner exception stack trace ---
   at Dicom.Network.DicomService.<TryCloseConnectionAsync>d__68.MoveNext() in C:\dev\GitHub\fo-dicom\fo-dicom\DICOM\Network\DicomService.cs:line 1471
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Dicom.Network.DicomService.<SendNextPDUAsync>d__60.MoveNext() in C:\dev\GitHub\fo-dicom\fo-dicom\DICOM\Network\DicomService.cs:line 373
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Dicom.Network.DicomService.PDataTFStream.<WritePDUAsync>d__13.MoveNext() in C:\dev\GitHub\fo-dicom\fo-dicom\DICOM\Network\DicomService.cs:line 1739
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Dicom.Network.DicomService.PDataTFStream.<CreatePDVAsync>d__12.MoveNext() in C:\dev\GitHub\fo-dicom\fo-dicom\DICOM\Network\DicomService.cs:line 1712
2022-02-09 11:39:49,983 - [50] - ERROR - Dicom - Exception writing data to PDV: System.IO.IOException: Unable to write data to the transport connection: An existing connection was forcibly closed by the remote host. ---> System.Net.Sockets.SocketException: An existing connection was forcibly closed by the remote host
   at System.Net.Sockets.Socket.EndSend(IAsyncResult asyncResult)
   at System.Net.Sockets.NetworkStream.EndWrite(IAsyncResult asyncResult)
   --- End of inner exception stack trace ---
   at Dicom.Network.DicomService.<TryCloseConnectionAsync>d__68.MoveNext() in C:\dev\GitHub\fo-dicom\fo-dicom\DICOM\Network\DicomService.cs:line 1471
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Dicom.Network.DicomService.<SendNextPDUAsync>d__60.MoveNext() in C:\dev\GitHub\fo-dicom\fo-dicom\DICOM\Network\DicomService.cs:line 373
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Dicom.Network.DicomService.PDataTFStream.<WritePDUAsync>d__13.MoveNext() in C:\dev\GitHub\fo-dicom\fo-dicom\DICOM\Network\DicomService.cs:line 1739
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Dicom.Network.DicomService.PDataTFStream.<CreatePDVAsync>d__12.MoveNext() in C:\dev\GitHub\fo-dicom\fo-dicom\DICOM\Network\DicomService.cs:line 1718
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Dicom.Network.DicomService.PDataTFStream.<WriteAsync>d__30.MoveNext() in C:\dev\GitHub\fo-dicom\fo-dicom\DICOM\Network\DicomService.cs:line 1819
2022-02-09 11:39:49,986 - [50] - ERROR - Dicom - Exception sending DIMSE: System.IO.IOException: Unable to write data to the transport connection: An existing connection was forcibly closed by the remote host. ---> System.Net.Sockets.SocketException: An existing connection was forcibly closed by the remote host
   at System.Net.Sockets.Socket.EndSend(IAsyncResult asyncResult)
   at System.Net.Sockets.NetworkStream.EndWrite(IAsyncResult asyncResult)
   --- End of inner exception stack trace ---
   at Dicom.DicomDatasetWalker.<DoWalkAsync>d__11.MoveNext() in C:\dev\GitHub\fo-dicom\fo-dicom\DICOM\DicomDatasetWalker.cs:line 379
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Dicom.Network.DicomService.<DoSendMessageAsync>d__66.MoveNext() in C:\dev\GitHub\fo-dicom\fo-dicom\DICOM\Network\DicomService.cs:line 1351
2022-02-09 11:39:49,986 - [50] - INFO - Dicom - PACSAE -> C-Store request [12]
```

**To Reproduce**
Very difficult

**Expected behavior**
SendAsync() should return or throw exception.

**Environment**
Fellow Oak DICOM version: 4.0.8
OS: Windows 10 IoT Enterprise LTSB 2016
Platform: .NET Framework 4.5.2

This is almost identical to issue #1054
#1058 has not resolved the issue.
Passing a Cancellation Token to SendAsync does not resolve the issue.

**Comment by amoerie** [comment]
Hi and thanks for providing this info. I'm afraid this is still somewhat of a known issue, although I disagree with your last statement:

> Passing a Cancellation Token to SendAsync does not resolve the issue.

If you make a cancellation token that triggers after 10 minutes (for example), DicomClient will properly break down any associations / connections it had created and stop hanging. I do concede that this is not very "optimal", but at the very least it allows you to provide an escape hatch.

If you have found a scenario where cancelling the token is not respected by DicomClient, please log a separate issue for that!

In the meantime, I have added logging in some places. Sometimes, a few of our unit tests fail in CI, which seems directly related to this. I'll try to follow up on this when it occurs again, but I'm afraid that today I don't have enough information to fix this.

If you would find a reproduction scenario, or - even better - a unit test that demonstrates the issue, I could be of immediate help.

**Comment by saydo17** [comment]
Unfortunately, I haven't been able to come up with a 100% reliable unit tests for this.
The test only fails if Async Operations are > 1.  When it does lock up, the DicomService is stuck waiting on the _pduQueueWatcher.

**Comment by amoerie** [comment]
Thanks for sharing what you know. This issue has been haunting us for a while, and since it survived the DicomClient refactorings I'm fairly confident the issue must be in the neighborhood of DicomService, but I haven't been able to pinpoint it more accurately yet. If only we had a consistent reproduction scenario, this would be a lot easier.

**Comment by amoerie** [comment]
@saydo17 

I may have found the root cause. This is the scenario that fits all of the symptoms:

1. You send DICOM request 1 to a DICOM server
2. That DICOM server goes offline/unreachable at some random moment
3. At the same time, you try to send DICOM request 2 to the same DICOM server. While thread A notices the disconnection and calls into DicomService.TryCloseConnection, thread B starts sending the second request. Let's follow thread B and assume thread A is very slow.
4. What if thread B manages to propagate all the way into DicomService:
  4.1. DicomService.SendRequestAsync
  4.2. DicomService.SendMessageAsync
  4.3. DicomService.SendNextMessageAsync
  4.4. DicomService.DoSendMessageAsync
  4.5. (inside DoSendMessageAsync) datasetWalker.WalkAsync(writer) which writes into a PDataTFStream
  4.6. PDataTFStream.WritePDUAsync
  4.7. DicomService.SendPDUAsync
  4.8. Here, thread B infinitely waits for _pduQueueWatcher

Inside the path to 4.8, thread B passes a few checks for whether we actually still have a connection or not, but if the race condition is unlucky enough to make it past all those checks and ends up at the _pduQueueWatcher, there are no remaining mechanisms that can wake it up when the _isDisconnectedFlag is finally set.

I think the solution should be twofold:
1. When disconnecting (inside DicomService.TryCloseConnection), also set the pduQueueWatcher flag to wake up any waiting threads. When they continue, they will immediately see the connection is gone and complete.
2. When waiting for the pduQueueWatcher flag, check for disconnection every minute if, for some reason, the pdu queue watcher flag is never set.

I've created a separate PR here: https://github.com/fo-dicom/fo-dicom/pull/1332

**Comment by amoerie** [comment]
See these commits for more info:

[`0b5d92a` (#1144)](https://github.com/fo-dicom/fo-dicom/pull/1144/commits/0b5d92a6eb76ed7fb8425cabec45929e51829d68)
[`71a85d2` (#1144)](https://github.com/fo-dicom/fo-dicom/pull/1144/commits/71a85d230cd15ef5c222f012154ff4fd5ac3f6a0)

**Comment by amoerie** [comment]
@saydo17 is it possible for you to upgrade to the v5 release line? Because currently the PR only applies a fix to the next 5.x release.

Technically we don't make new releases for the 4.x line anymore, except for security issues or critical bugs. 

We could probably backport this fix if need be, but it has to be worth the hassle.

**Comment by saydo17** [comment]
We will have to investigate updating to v5.  Currently we target .net 4.5.2.  As such, .net Standard 2.0 is not available to us.

**Comment by saydo17** [comment]
We have updated to v5.

**Comment by amoerie** [comment]
> We have updated to v5.

Excellent! The pull request is still open, waiting for feedback from @gofal 

We could already merge it anyway, but since he is the only one who can publish new versions on NuGet, that would be meaningless.
